### PR TITLE
Yarn: Move node_modules, package.json, and yarn.lock file to vendor

### PR DESCRIPTION
**Pull request by dhh**
And then provide a binstub for yarn that runs in that directory.

### CODE REVIEW

File: railties/lib/rails/generators/rails/app/templates/bin/yarn [hunk 1]
```diff
@@ -0,0 +1,2 @@
+VENDOR_PATH = File.expand_path('../vendor', __dir__)
+Dir.chdir(VENDOR_PATH) { exec "yarnpkg #{ARGV.join(" ")}" }
```
Comment (ianks) [inline railties/lib/rails/generators/rails/app/templates/bin/yarn]: Booting a whole ruby interpreter just to shell out to yarn seems unnecessary. This is especially painful for jruby folks.

File: railties/CHANGELOG.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
-*   Add Yarn support in new apps using --yarn option. This adds a package.json
+*   Add Yarn support in new apps using --yarn option. This adds a yarn binstop, vendor/package.json,
```
Comment (ianks) [inline railties/CHANGELOG.md]: binstub*

File: railties/lib/rails/generators/rails/app/templates/bin/yarn (hunk 1 above)
Comment (dhh) [inline railties/lib/rails/generators/rails/app/templates/bin/yarn]: Booting Ruby isn't time consuming with MRI. But happy to take a patch with another approach that works on all platforms (and thus doesn't rely on Bash or whatever).

Comment (ianks) [inline railties/lib/rails/generators/rails/app/templates/bin/yarn]: FWIW,  the `yarn` binstub relies on sh already.

**Comment by lemonmade** [comment]
I am really interested why this decision was made. It will make actually configuring any JavaScript tooling extremely difficult, and causes it to behave differently than it would in any other project that uses npm packages. As the simplest example, editor integration with a popular JavaScript linter like ESLint will not work anymore as they all rely on a "traditional" structure that has `node_modules` at the root, and that follows the basic node resolution algorithm.

At Shopify, we've had no trouble with `node_modules` at the root of the app (plus additional `node_modules` for separate bundles within the main application). It's been easier for new developers to understand where dependencies are coming from since it's just like any other npm-using project. If the only benefit is that the root directory is cleaned up a little bit, I don't see that as being worth the cost, particularly when other dependency information for the application (the Gemfile) is already stored at root.

**Comment by matthewd** [comment]
> extremely difficult

@lemonmade can't you just create a top level `node_modules` directory if that works better for you? :confused:

**Comment by dhh** [comment]
This isn't about whether node_modules + package.json + yarn.lock filling
into the root is "trouble" but whether it's a beautiful way to cohabit
app-focused JS within a Rails app. I don't think it is. As explored
earlier, I don't think that Gemfile* in the root is a good fit either. I'd
rather that both of those configurations lived in vendor/. And we'll work
to see if we can get there eventually too. But I'm not keen to compound the
problem with more vendor-configuration and install directories at our
already-busy root level.

We've found it pretty painless to configure both Yarn and Webpack to work
with this cohabitation structure. I'd be surprised if there weren't also
ways to make something like ESLint work as well.

Let's keep exploring.

**Comment by ianks** [comment]
Having worked extensively with frontend tooling, I very much agree with @lemonmade. Pretty much all files that are in traditionally in will not be easily discoverable by tooling. It will end up just causing headaches for users by breaking conventions that already exist to create a new convention. 

I think by default, `node_modules` and `yarn.lock` should be in the root directory. This way, the default experience will be painless as all node tooling can happily assume that executable will be in `node_modules/.bin`, etc. If someone wants move these to `vendor`, that should be fine; but IMO, don't cause integration pains for other users because you want a subjectively "cleaner" directory structure.

**Comment by dhh** [comment]
Appreciate the input and will keep it under advisement as we go forward.
This isn't set in stone, but for now we're going to try to make this work.

Also, beautiful code and structures may well be "subjective", but it's a
core value for Rails (see http://rubyonrails.org/doctrine/#beautiful-code
for more). So let's try to make this work. If we put in all the effort and
its still a bust in the end, we can take a different path from there.

**Comment by ianks** [comment]
As a counter-argument, convention over configuration is also a core value; which this implementation will invariably break. I do not think it is feasible to change the conventions of all the tooling around node to accommodate for this use case. 

Here is what will happen in my case:

1. My editor will look for a local version of `mocha`,  `tsc (typescript`, or `sass-lint` inside of node_modules in the project root directory.
2. When it does not find it, it will fall back to the global version of these commands
3. These global versions will either:
    a. Not exist (best case)
    b. Exist, but potentially be a different version that what I have specified in my yarn.lock (worst case). Now, my editor is telling me things that are potentially wrong or outdated.
4. This will cause me to waste my time attempting to reconfigure things/convincing the upstream editor to apply a patch which will fix my use case. None of this sounds like a pleasant experience.

**Comment by dhh** [comment]
I understand That's Just The Way Things Are but I always mentally
compliment that with "for the moment" and "doing things as we do". The
birth of this project was to push back against those things. Work with
others to change things and integrate them in ways for mutual benefit. I'd
like to keep trying to do that here.

**Comment by nathanmarks** [comment]
The issue here is that in order not to have a `node_modules` folder at the root (which doesn't get checked in anyways), conventions are being broken across the board.

If we're using a framework like rails to begin with, it stands to reason that we enjoy the comfortable conventions that the tools we use offer. I see no reason why rails has to break these JS conventions in order to accommodate JS, a critical component of web development.

1. As @ianks mentioned, it'll mess with editor plugins and other related tools. Everything will need custom configuration (or potentially patches) just to work with Rails. Isn't the whole point of rails convention over configuration?
2. Take a simple feature such as Heroku buildpack inference, or CI inference -- they work on the basis that there is a `package.json` present in the project root. Point being... these are conventions that are respected across the board. All of a sudden, further configuration here is required to get started for the simplest use cases.
3. Historically, (and not just in rails), `/vendor` folders were/are used to place 3rd party dependencies that are used in the application itself, not the dev tooling used to build the project. In fact, before package management was around for JS (bower, npm, etc), using a `/vendor` folder was also very common for non-rails projects.
4. Unless I'm missing something, the only way to make node module resolution work for the `/vendor/node_modules` folder is either by:
  a. Using `$NODE_PATH` which is not recommended unless there's absolutely no other way. (Especially if it means that node module resolution is changed out of the box for anyone using rails for the sake of a single folder that isn't checked into version control... we're not talking about moving the assets folder to the root or anything crazy)
  b. Monkey patching node's `Module` object at runtime
  c. Using relative requires to get things from the `node_modules` folder...

**Comment by dhh** [comment]
@nathanmarks Thanks for your considerations. Good to have all this spelled out when we evaluate the current approach a bit further down the line. We're indeed using $NODE_PATH and I currently consider that to be a fully justified use. But as stated earlier: This isn't set in stone. We're going to try it for real and make it work, but if we can't, we'll reconsider. Let's resume this discussion in a month or two. In the mean time, eager to have everyone's help improving things.

https://github.com/rails/rails/pull/27288 is now available.

**Comment by olivierlacan** [comment]
For what it's worth, it might be possible to add the option to set the `--modules-folder` inside of a `.yarnrc` file (so the Rails install directory works without a binstub). 

The maintainers [seem open to it](https://github.com/yarnpkg/yarn/issues/2782#issuecomment-282896097).

**Comment by gsamokovarov** [comment]
In the end, David set it to the root Rails application folder in https://github.com/rails/rails/pull/28093.

**Comment by boompig** [comment]
Is there any documentation on whether `modules-folder` can be set in `.yarnrc`? All the google searches for me result in github issues asking for this feature.

**Pull request merged**